src/test/coins_tests.cpp

```diff
@@ -562,20 +564,57 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+
+    bool operator==(const CoinEntry& o) const = default;
+    friend std::ostream& operator<<(std::ostream& os, const CoinEntry& e) { return os << e.value << e.state; }
+
+    constexpr bool IsDirtyFresh() const { return state == State::DIRTY_FRESH; }
+    constexpr bool IsDirty() const { return state == State::DIRTY || IsDirtyFresh(); }
+    constexpr bool IsFresh() const { return state == State::FRESH || IsDirtyFresh(); }
+
+    static constexpr CoinEntry::State ToState(bool is_dirty, bool is_fresh) {
+        if (is_dirty && is_fresh) return CoinEntry::State::DIRTY_FRESH;
+        if (is_dirty) return CoinEntry::State::DIRTY;
+        if (is_fresh) return CoinEntry::State::FRESH;
+        return CoinEntry::State::CLEAN;
+    }
+};
+
 const static COutPoint OUTPOINT;
-const static CAmount SPENT = -1;
-const static CAmount ABSENT = -2;
-const static CAmount FAIL = -3;
-const static CAmount VALUE1 = 100;
-const static CAmount VALUE2 = 200;
-const static CAmount VALUE3 = 300;
-const static char DIRTY = CCoinsCacheEntry::DIRTY;
-const static char FRESH = CCoinsCacheEntry::FRESH;
-const static char NO_ENTRY = -1;
-
-const static auto FLAGS = {char(0), FRESH, DIRTY, char(DIRTY | FRESH)};
-const static auto CLEAN_FLAGS = {char(0), FRESH};
-const static auto ABSENT_FLAGS = {NO_ENTRY};
+constexpr CAmount SPENT = -1;
+constexpr CAmount ABSENT = -2;
+constexpr CAmount VALUE1 = 100;
+constexpr CAmount VALUE2 = 200;
+constexpr CAmount VALUE3 = 300;
+
+using MaybeCoin = std::optional<CoinEntry>;
+using CoinOrError = std::variant<MaybeCoin, std::string>;
+
+constexpr MaybeCoin MISSING            = std::nullopt;
+constexpr MaybeCoin SPENT_DIRTY        = CoinEntry{SPENT, CoinEntry::State::DIRTY};
+constexpr MaybeCoin SPENT_DIRTY_FRESH  = CoinEntry{SPENT, CoinEntry::State::DIRTY_FRESH};
+constexpr MaybeCoin SPENT_FRESH        = CoinEntry{SPENT, CoinEntry::State::FRESH};
+constexpr MaybeCoin SPENT_CLEAN        = CoinEntry{SPENT, CoinEntry::State::CLEAN};
+constexpr MaybeCoin VALUE1_DIRTY       = CoinEntry{VALUE1, CoinEntry::State::DIRTY};
+constexpr MaybeCoin VALUE1_DIRTY_FRESH = CoinEntry{VALUE1, CoinEntry::State::DIRTY_FRESH};
+constexpr MaybeCoin VALUE1_FRESH       = CoinEntry{VALUE1, CoinEntry::State::FRESH};
+constexpr MaybeCoin VALUE1_CLEAN       = CoinEntry{VALUE1, CoinEntry::State::CLEAN};
+constexpr MaybeCoin VALUE2_DIRTY       = CoinEntry{VALUE2, CoinEntry::State::DIRTY};
+constexpr MaybeCoin VALUE2_DIRTY_FRESH = CoinEntry{VALUE2, CoinEntry::State::DIRTY_FRESH};
+constexpr MaybeCoin VALUE2_FRESH       = CoinEntry{VALUE2, CoinEntry::State::FRESH};
+constexpr MaybeCoin VALUE2_CLEAN       = CoinEntry{VALUE2, CoinEntry::State::CLEAN};
+constexpr MaybeCoin VALUE3_DIRTY       = CoinEntry{VALUE3, CoinEntry::State::DIRTY};
+constexpr MaybeCoin VALUE3_DIRTY_FRESH = CoinEntry{VALUE3, CoinEntry::State::DIRTY_FRESH};
+
+const static std::string EX_OVERWRITE_UNSPENT = "Attempted to overwrite an unspent coin (when possible_overwrite is false)";
```

Comment:
Since #30921, we can now make these `constexpr string_view`.

Reply:
Made it a `constexpr auto` (I don't want non-owning fields) https://github.com/bitcoin/bitcoin/compare/1dd79a32c9c6e0c6cf4ecef00b56972574babced..cb9be6729c4e27bf2d6d703c0d454a22cbcdb6e1